docs/integrations/Instructions.md

```diff
@@ -0,0 +1,160 @@
+act on invalid data, or send to itself, it will drop the entire instruction.
+
+### Serialization
+
+  - Numbers are exclusively unsigned and encoded as compact integers under
+SCALE. If omitted, `0`.
+  - Enums are prefixed by an ordinal byte of their type, followed by their
+actual values.
+  - Vectors are prefixed by their length. If omitted, `vec![]`.
+  - Instruction fields are numbered and sequentially encoded, each prefixed by
+an ordinal byte. All other fields are sequentially encoded with no markers.
+
+Certain fields may be omitted depending on the network in question.
+
+### In Instructions
+
+  - `origin` (Address): Address from the network of origin which sent funds in.
+  - `target` (Address): The ink! contract to transfer the incoming funds to.
+  - `data`   (Vec<u8>): The data to call the target with.
+
+Networks may automatically provide `origin`. If they do, the instruction may
+still provide `origin`, overriding the automatically provided value. If no
+`origin` is provided, the instruction is dropped.
+
+Upon receiving funds, the respective Serai Asset contract is called, minting the
+appropriate amount of coins, and transferring them to the specified target with
+the attached data.
+
+If the transaction fails, funds are scheduled to be returned to `origin`.
+
+### Out Instructions
+
+  - `destination` (Enum { Native(Address), Serai(Address) }): Address to receive
+funds to.
+  - `data`        (Vec<u8>):                                  The data to call
+the target with.
+
+If the network is Serai, this is a transfer. Else, it's a withdrawal to the
+specified address with the specified data. Asset contracts perform no validation
+on these fields.
+
+### Shorthand
+
+All In Instructions are encoded as Shorthand. Shorthand is an enum which expands
+to an In Instruction.
+
+##### Raw
+
+Raw Shorthand encodes a raw In Instruction with no further processing.
+
+##### Swap
+
+  - `origin`  (Option<Address>): In Instruction's `origin`.
+  - `coin`    (Coin):            Coin to swap funds for.
+  - `minimum` (Amount):          Minimum amount of `coin` to receive.
+  - `out`     (Out Instruction): Final destination for funds.
+
+which expands to:
+
+```
+In Instruction {
+  origin,
+  target: Router,
+  data:   swap(Incoming Asset, out, minimum)
+}
+```
+
+where `swap` is a function which:
+
+  1) Swaps the incoming funds for SRI.
+  2) Swaps the SRI for `coin`.
+  3) Checks the amount of `coin` received is greater than `minimum`.
+  4) Executes `out` with the amount of `coin` received.
+
+For a Bitcoin to Monero swap, Swap Shorthand is expected to generally take:
+
+  - 1 byte to identify as Swap.
+  - 1 byte to not override `origin`.
+  - 1 byte for `coin`.
+  - 4 bytes for `minimum`.
+  - 1 byte for `out`'s `destination` field label.
+  - 1 byte for `out`'s `destination`'s ordinal byte.
+  - 65 bytes for `out`'s `destination`'s address.
+
+Or 74 bytes.
+
+##### Add Liquidity
```

Comment:
Ideally, Add Liq would solely be from Serai. The issue is the Monero UX.

Monero can:

- Swap and add (1 XMR TX, 108 bytes)
- Swap for Serai + gas, wrap, add liquidity (1 XMR TX, 107 bytes, wait 20 minutes, 1 XMR TX, ~98 bytes, 1 Serai TX)
- Swap for Serai while also wrapping (1 XMR TX, 111 bytes, 1 Serai TX)

So I believe we have to accept this exposure and 108 bytes on Monero, not including the 1 byte label included on Monero (and tx.extra VarInt length size increase, +1 byte there too).